klayout/drc/run_drc.py

```diff
@@ -0,0 +1,709 @@
+    --run_dir=<run_dir_path>            Run directory to save all the results [default: pwd]
+    --thr=<thr>                         The number of threads used in run.
+    --run_mode=<run_mode>               Select klayout mode Allowed modes (flat , deep, tiling). [default: flat]
+    --no_feol                           Turn off FEOL rules from running.
+    --no_beol                           Turn off BEOL rules from running.
+    --connectivity                      Turn on connectivity rules.
+    --density                           Turn on Density rules.
+    --density_only                      Turn on Density rules only.
+    --antenna                           Turn on Antenna checks.
+    --antenna_only                      Turn on Antenna checks only.
+    --no_offgrid                        Turn off OFFGRID checking rules.
+    --verbose                           Detailed rule execution log for debugging.
+"""
+
+from docopt import docopt
+import os
+import xml.etree.ElementTree as ET
+import logging
+import klayout.db
+import glob
+from datetime import datetime
+from subprocess import check_call
+import shutil
+import concurrent.futures
+import traceback
+
+
+def get_rules_with_violations(results_database):
+    """
+    This function will find all the rules that has violated in a database.
+
+    Parameters
+    ----------
+    results_database : string or Path object
+        Path string to the results file
+
+    Returns
+    -------
+    set
+        A set that contains all rules in the database with violations
+    """
+
+    mytree = ET.parse(results_database)
+    myroot = mytree.getroot()
+
+    all_violating_rules = set()
+
+    for z in myroot[7]:  # myroot[7] : List rules with viloations
+        all_violating_rules.add(f"{z[1].text}".replace("'", ""))
+
+    return all_violating_rules
+
+
+def check_drc_results(results_db_files: list):
+    """
+    check_drc_results Checks the results db generated from run and report at the end if the DRC run failed or passed.
+    This function will exit with 1 if there are violations.
+
+    Parameters
+    ----------
+    results_db_files : list
+        A list of strings that represent paths to results databases of all the DRC runs.
+    """
+
+    if len(results_db_files) < 1:
+        logging.error("Klayout did not generate any rdb results. Please check run logs")
+        exit(1)
+
+    full_violating_rules = set()
+
+    for f in results_db_files:
+        violating_rules = get_rules_with_violations(f)
+        full_violating_rules.update(violating_rules)
+
+    if len(full_violating_rules) > 0:
+        logging.error("Klayout DRC run is not clean.")
+        logging.error("Violated rules are : {}\n".format(str(full_violating_rules)))
+        exit(1)
+    else:
+        logging.info("Klayout DRC run is clean. GDS has no DRC violations.")
+
+
+def get_results(rule_deck, rules, lyrdb, type):
```

Comment:
the name suggest this should return something.

Reply:
removed this function, no longer used and replaced by 'check_drc_results'